### PR TITLE
Merge development_3.0 into development

### PR DESCRIPTION
**Pull request by daverodgman**
This supersedes #4399

Merge development_3.0 into development, plus an empty commit to address sign-offs to a few commits where this was missing.

### CODE REVIEW

**Review by yanesca — APPROVED** [review]
Looks good to me. I've checked that:

- the target branch is correct
- the content is identical to #4399
- the CI is passing (except the API-ABI check and DCO check which fail as expected)
- the additional commit covers all commits missing sign-off

**Review by gilles-peskine-arm — COMMENTED** [review]
This is a correct merge of `development_3.0` plus one acceptable commit, effectively bringing the `development` branch to containing the 3.0 work.

The extra commit declares a sign-off for 3 commits that were missing one.

My approval is conditional on:
* CI passing (other than the DCO check on the 3 listed commits).
* The existence of public guidance as to when to target `development_2.x` from now on.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
On second thoughts my [previous review](https://github.com/ARMmbed/mbedtls/pull/4407#pullrequestreview-643388566) was incomplete. This is a correct merge, but some things are missing.

The documentation should make it clear that this is an unstable branch of Mbed TLS. This was already missing in the `development_3.0` branch, and becomes a lot more important now that this is the default branch.

At a minimum:
* `README.md` needs a mention of the instability near the top, with a pointer to the 2.x branch.
* `BRANCHES.md` needs updating.
* `ChangeLog` should have an open entry stating that there is ongoing work for the next revision (`= Mbed TLS 3.0.0 branch released 2021-xx-xx` — xx is what we use for yet-unknown numbers in unfinished versions, and let's take this opportunity to apply the trademark typography change mbed→Mbed). I suggest running `assemble_changelog.py` on 3.0.

@yanesca Do you see other necessary documentation changes?

File: README.md [hunk 1]
```diff
@@ -5,6 +5,13 @@ Mbed TLS is a C library that implements cryptographic primitives, X.509 certific
 
 Mbed TLS includes a reference implementation of the [PSA Cryptography API](#psa-cryptography-api). This is currently a preview for evaluation purposes only.
 
+Stablility
+----------
+
+The [`development_2.x`] branch will contain work for the next Mbed TLS 2.x release, and will retain a stable API.
+
+The [`development`] branch is where work focused on the up-coming Mbed TLS 3.0 release takes place - this means that (in addition to everything in [`development_2.x`]) it will introduce API incompatibilities and is not recommended for users. New features will normally land here.
```
Comment (gilles-peskine-arm) [inline README.md:13]: Here's my suggested text:
```suggestion
**Warning: this branch of Mbed TLS currently has an unstable API.** It is where work is happening on the next major release of Mbed TLS. Until Mbed TLS 3.0 is released, if you need a stable API, please use the branch [`development_2.x`](https://github.com/ARMmbed/mbedtls/tree/development_2.x) instead.
```

File: BRANCHES.md [hunk 2]
```diff
@@ -6,8 +6,10 @@ At any point in time, we have a number of maintained branches consisting of:
   this always contains the latest release, including all publicly available
   security fixes.
 - The [`development`](https://github.com/ARMmbed/mbedtls/tree/development) branch:
-  this is where new features land,
-  as well as bug fixes and security fixes.
+  this is where new features normally land, as well as bug fixes and security
+  fixes.
+- The [`development_2.x`] branch:
```
Comment (gilles-peskine-arm) [inline BRANCHES.md:11]: Missing link.

File: BRANCHES.md [hunk 3]
```diff
@@ -6,8 +6,10 @@ At any point in time, we have a number of maintained branches consisting of:
...
+  this is where features and bug-fixes destined for 2.x land.
```
Comment (gilles-peskine-arm) [inline BRANCHES.md:12]: Here's my suggested text:
```suggestion
   this is where the next major version of Mbed TLS (version 3.0 is being
   prepared. It has API changes that make it incompatible with Mbed TLS 2.x,
   as well as all the new features and bug fixes and security fixes.
- The [`development_2.x`](https://github.com/ARMmbed/mbedtls/tree/development_2.x) branch:
   this branch retains the API of Mbed TLS 2.x, and has a subset of the
   features added after Mbed TLS 2.26.0 and bug fixes and security fixes.
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

**Comment by daverodgman** [comment]
@gilles-peskine-arm thanks, I've updated in line with your suggestions. I haven't replaced x.x.x with 3.0.0 - this is generated via assemble_changelog.py so I'd either have to edit the autogenerated output, which seems wrong/confusing, or assemble_changelog.py itself, which is out of scope for this PR.

I've also corrected a wrongly named Changelog file to ensure it gets picked up by the Changelog script.

**Comment by daverodgman** [comment]
As per discussion, updated Changelog manually to adjust the version number, date and capitalisation.

**Review by yanesca — COMMENTED** [review]
Found a couple of typos, other than that it looks good to me.

**Review by yanesca — APPROVED** [review]
LGTM. (Assuming that the CI passes and Gilles's requests have been addressed.)

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Pull request merged**